src/runtime/policy/policy.ts

```diff
@@ -0,0 +1,270 @@
+    const annotationRefs = buildAnnotationRefs(node.annotationRefs);
+
+    checkNamesAreUnique(node.targets.map(target => ({name: target.schemaName})));
+    const targets = node.targets.map(target => PolicyTarget.fromAstNode(target, buildAnnotationRefs));
+
+    checkNamesAreUnique(node.configs);
+    const configs = node.configs.map(config => PolicyConfig.fromAstNode(config));
+
+    return new Policy(node.name, targets, configs, annotationRefs);
+  }
+}
+
+class PolicyTarget {
+  constructor(
+      readonly schemaName: string,
+      readonly fields: PolicyField[],
+      private readonly annotationRefs: AnnotationRef[]) {
+    this.validateRetentions();
+    // TODO(b/157605585): Validate maxAge annotation.
+  }
+
+  get retentions() {
+    return this.annotationRefs
+        .filter(annotation => annotation.name === 'allowedRetention')
+        .map(annotation => ({
+          medium: PolicyRetentionMedium[annotation.params['medium'] as string],
+          encryptionRequired: annotation.params['encryption'] as boolean,
+        }));
+  }
+
+  get annotations() {
+    return this.annotationRefs.filter(annotation => annotation.name !== 'allowedRetention');
+  }
+
+  private validateRetentions() {
+    // Check retention annotations are valid.
+    const retentionMediums: Set<PolicyRetentionMedium> = new Set();
+    this.annotationRefs.filter(annotation => annotation.name === 'allowedRetention').forEach(annotation => {
+      const mediumParam = annotation.params['medium'];
+      checkValueInEnum(mediumParam, PolicyRetentionMedium);
+      const medium = mediumParam as PolicyRetentionMedium;
+      if (retentionMediums.has(medium)) {
+        throw new Error(`@allowedRetention has already been defined for ${medium}.`);
+      }
+      retentionMediums.add(medium);
+    });
+  }
+
+  toManifestString(): string {
+    return [
+      ...this.annotationRefs.map(annotation => annotation.toString()),
+      `from ${this.schemaName} access {`,
+      ...flatMap(this.fields, field => indentLines(field.toManifestString())),
+      '}',
+    ].join('\n');
+  }
+
+  static fromAstNode(
+      node: AstNode.PolicyTarget,
+      buildAnnotationRefs: (ref: AstNode.AnnotationRef[]) => AnnotationRef[]): PolicyTarget {
+    const annotationRefs = buildAnnotationRefs(node.annotationRefs);
+
+    // Convert fields.
+    checkNamesAreUnique(node.fields);
+    const fields = node.fields.map(field => PolicyField.fromAstNode(field, buildAnnotationRefs));
+
+    return new PolicyTarget(node.schemaName, fields, annotationRefs);
+  }
+}
+
+class PolicyField {
+  constructor(
+      readonly name: string,
+      readonly subfields: PolicyField[],
+      private readonly annotationRefs: AnnotationRef[]) {
+    this.validateAllowedUsages();
+    // TODO(b/157605585): Validate field structure against Type.
+  }
+
+  get allowedUsages() {
+    // TODO(b/157605585): Handle "raw" label separately?
+    return this.annotationRefs
```

Comment:
maybe instead of doing it in a getter, initialize an allowedUsages data member in the ctor.
this getter and the validateAllowedUsages method have duplicate code, seems suboptimal

Reply:
Done